pydantic_ai_slim/pydantic_ai/ag_ui.py

```diff
@@ -185,320 +189,349 @@ def __init__(
+        deps: Optional dependencies to use for this run.
+        model_settings: Optional settings to use for this model's request.
+        usage_limits: Optional limits on model request count or token usage.
+        usage: Optional usage to start with, useful for resuming a conversation or agents used in tools.
+        infer_name: Whether to try to infer the agent name from the call frame if it's not set.
+        toolsets: Optional additional toolsets for this run.
+
+    Returns:
+        A streaming Starlette response with AG-UI protocol events.
     """
+    accept = request.headers.get('accept', SSE_CONTENT_TYPE)
+    try:
+        input_data = RunAgentInput.model_validate(await request.json())
+    except ValidationError as e:  # pragma: no cover
+        return Response(
+            content=json.dumps(e.json()),
+            media_type='application/json',
+            status_code=HTTPStatus.UNPROCESSABLE_ENTITY,
+        )
 
-    agent: Agent[AgentDepsT, OutputDataT] = field(repr=False)
-
-    async def run(
-        self,
-        run_input: RunAgentInput,
-        accept: str = SSE_CONTENT_TYPE,
-        *,
-        output_type: OutputSpec[RunOutputDataT] | None = None,
-        model: Model | KnownModelName | str | None = None,
-        deps: AgentDepsT = None,
-        model_settings: ModelSettings | None = None,
-        usage_limits: UsageLimits | None = None,
-        usage: Usage | None = None,
-        infer_name: bool = True,
-        toolsets: Sequence[AbstractToolset[AgentDepsT]] | None = None,
-    ) -> AsyncGenerator[str, None]:
-        """Run the agent with streaming response using AG-UI protocol events.
+    return StreamingResponse(
+        run_ag_ui(
+            agent,
+            input_data,
+            accept,
+            output_type=output_type,
+            model=model,
+            deps=deps,
+            model_settings=model_settings,
+            usage_limits=usage_limits,
+            usage=usage,
+            infer_name=infer_name,
+            toolsets=toolsets,
+        ),
+        media_type=accept,
+    )
 
-        The first two arguments are specific to `Adapter` the rest map directly to the `Agent.iter` method.
 
-        Args:
-            run_input: The AG-UI run input containing thread_id, run_id, messages, etc.
-            accept: The accept header value for the run.
+async def run_ag_ui(
+    agent: Agent[AgentDepsT, Any],
+    run_input: RunAgentInput,
+    accept: str = SSE_CONTENT_TYPE,
+    *,
+    output_type: OutputSpec[Any] | None = None,
+    model: Model | KnownModelName | str | None = None,
+    deps: AgentDepsT = None,
+    model_settings: ModelSettings | None = None,
+    usage_limits: UsageLimits | None = None,
+    usage: Usage | None = None,
+    infer_name: bool = True,
+    toolsets: Sequence[AbstractToolset[AgentDepsT]] | None = None,
+) -> AsyncGenerator[str, None]:
```

Comment:
I'm thinking we should somehow be returning the agent run result object as well, in case you want to persist the message history or something, but there's no obvious way to do that in a generator that can only yield, not return...

Reply:
CopilotKit already handles its own thread persistence, so this is probably not necessary: https://docs.copilotkit.ai/pydantic-ai/persistence/loading-message-history

Reply:
Since the events are always going to be streamed straight to the frontend by having the endpoint function return a `StreamingResponse` wrapping the return value of this function, and since the run result object is only usable after streaming is complete, there really is no way you could use it anyway. So I think we're OK losing it -- all of the information will be in the streamed messages anyway.

Reply:
Starlette middleware was what I was thinking